flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/source/PubSubSource.java

```diff
@@ -0,0 +1,304 @@
+            Preconditions.checkNotNull(subscriptionName);
+            this.subscriptionName = subscriptionName;
+            return this;
+        }
+
+        public PubSubSourceBuilder<OUT> withCredentials(Credentials credentials) {
+            this.credentials = credentials;
+            return this;
+        }
+
+        public PubSubSourceBuilder<OUT> withPubSubSubscriberFactory(
+                PubSubSubscriberFactory pubSubSubscriberFactory) {
+            this.pubSubSubscriberFactory = pubSubSubscriberFactory;
+            return this;
+        }
+
+        public PubSubSourceBuilder<OUT> withPubSubSubscriberFactory(
+                int maxMessagesPerPull, Duration perRequestTimeout, int retries) {
+            this.pubSubSubscriberFactory =
+                    new DefaultPubSubSubscriberFactory(
+                            ProjectSubscriptionName.format(projectName, subscriptionName),
+                            retries,
+                            perRequestTimeout,
+                            maxMessagesPerPull);
+            return this;
+        }
+
+        public PubSubSourceBuilder setProps(Properties props) {
+            this.props = props;
+            return this;
+        }
+
+        public PubSubSource<OUT> build() throws IOException {
+            if (credentials == null) {
+                credentials = defaultCredentialsProviderBuilder().build().getCredentials();
+            }
+
+            if (pubSubSubscriberFactory == null) {
+                pubSubSubscriberFactory =
+                        new DefaultPubSubSubscriberFactory(
+                                ProjectSubscriptionName.format(projectName, subscriptionName),
+                                3,
+                                Duration.ofSeconds(15),
+                                100);
+            }
+
+            return new PubSubSource(
+                    deserializationSchema, pubSubSubscriberFactory, props, credentials);
+        }
+    }
+
+    /** Part of {@link PubSubSourceBuilder} to set required fields. */
+    public static class DeserializationSchemaBuilder {
+        /**
+         * Set the DeserializationSchema used to deserialize incoming PubSubMessages. If you want
+         * access to meta data of a PubSubMessage use the overloaded
+         * withDeserializationSchema({@link PubSubDeserializationSchema}) method instead.
+         */
+        public <OUT> ProjectNameBuilder<OUT> withDeserializationSchema(
+                DeserializationSchema<OUT> deserializationSchema) {
+            return new PubSubSourceBuilder<>(deserializationSchema);
+        }
+
+        /** Set the DeserializationSchema used to deserialize incoming PubSubMessages. */
+        public <OUT> ProjectNameBuilder<OUT> withDeserializationSchema(
+                PubSubDeserializationSchema<OUT> deserializationSchema) {
+            return new PubSubSourceBuilder<>(deserializationSchema);
```

Comment:
I do no think this is a good way to expose it to the user. IMO you have to define your own deserialization Schema which is facilitated in your source and only let users implement this one.

Reply:
What do you mean exactly? That it should be forbidden to use (through the `DeserializationSchemaWrapper`) a schema that's not inheriting from `PubSubDeserializationSchema`?